Add file_reviews.linter_name

jbusser opened the pull request:
[Related Github issue](https://github.com/houndci/hound/issues/1101)

Generally speaking, for any given language, Hound supports only one linter. For
example, Rubocop will handle all `.rb` files; SCSS-lint will lint `.scss` files
and so on.

And then there's JavaScript.

For JavaScript, Hound supports three linters: JShint, JSCS, and ESlint,
each of which can be individually enabled or disabled. If both JShint and
JSCS are enabled, however, Hound code reviews will never complete for a given
build, and this is because `CompleteFileReivew#file_review` will only look up
a `FileReview` by its `filename` attribute, and when multiple linters are
working on a single file, that is not enough information. The second through nth
`FileReivew` for that file never gets found, and its `completed_at` never gets
updated. When that happens, `Build#complete?` always returns false.

This change creates a migration that adds a new column to the `file_reviews`
table, which will store the name of the linter assigned to a given file review.
The corresponding `FileReview` model will acquire a model-level validation to
ensure the presence of this new linter-name attribute.

It then updates `Linter::Base#enqueue_job` to add the linter_name attribute to
the hash of review-job attributes so that linters can pick up this new
attribute. When their reviews are complete, the linters can be configured to
pass the linter name back to Hound, along with any errors and the filename, just
as before.

The last part of this solution involves updating
`CompleteFileReview#file_review` to anticipate an optional `linter_name`
attribute in its dequeued job attributes, which it will use to perform a more
precise search of `FileReview` objects in the relevant build. In order to avoid
a large coordinated upgrade effort, this method will fall back to the previous
locator logic, which only operates on the job's `filename` attribute.

teoljungberg commented on app/models/linter/jscs.rb:
```diff
@@ -1,5 +1,7 @@
 module Linter
   class Jscs < Base
+    include JobWithLinterName
```
Could we create a class out of this instead? To use composition instead of inheritance?

jbusser replied:
Embarrassed to say I don't think I'm following you here. I thought my including a module was following composition over inheritance. What do you have in mind?

teoljungberg replied:
Including modules is basically the same as inheriting from a class.

I would imagine a class along the lines off to `JobAttributesFor.call(commit_file)`.

jbusser replied:
Rather than doing that, what if I went ahead and changed `Linter::Base#build_review_job_attributes` to supply the new linter_name attribute? The change is additive and therefore won't affect any linters, but it will be present in the payload if/when those linters ever need it.

Changing `Linter::Base#build_review_job_attributes` was my original approach but then saw it caused every linter's test to fail because their Resque payloads had changed, but perhaps it's worth the trouble? Or do you think a separate class is still a better design?

teoljungberg replied:
I would add it to `Linter::Base#build_review_job_attributes` and leave it at that.

We'll extract a new class when we need one.

gylaz replied:
:+1: on adding it to the `Linter::Base#build_review_job_attributes` if it doesn't break any of the services that don't support that argument.

However, say you wanted to only override it for a few linters, you could do:
```
module Linter
  class Jscs < Base
    def build_review_job_attributes(commit_file)
      super.merge(linter_name: "foo")
    end
  end
end
```

teoljungberg commented on app/models/linter/ruby.rb:
```diff
@@ -15,7 +15,9 @@ def file_included?(commit_file)
-      FileReview.create!(filename: commit_file.filename) do |file_review|
+      FileReview.create!(
+        filename: commit_file.filename,
+        linter_name: name) do |file_review|
```
This formatting is somewhat against how we have done things in Hound.

We usually format these expressions as:

```ruby
FileReview.create!(
  filename: commit_file.filename,
  linter_name: name,
) do |file_review|
# ...
end

```

jbusser replied:
No problem at all.

teoljungberg commented on app/services/complete_file_review.rb:
```diff
@@ -47,8 +47,10 @@ def build
-      find_by!(filename: attributes.fetch("filename"))
+    @file_review ||= (build.file_reviews.find_by(
+      filename: attributes.fetch("filename"),
+      linter_name: attributes["linter_name"]) ||
```
Same as above.

gylaz replied:
Can we extract this first part into a method to clean this up?

teoljungberg commented on db/migrate/20160314122828_add_linter_name_to_file_reviews.rb:
```diff
@@ -0,0 +1,5 @@
+class AddLinterNameToFileReviews < ActiveRecord::Migration
+  def change
+    add_column :file_reviews, :linter_name, :string
```
Should this have a default value, and `null: false`?

jbusser replied:
Sure. I could do that. I will remove the model-level validation then as well.

jbusser replied:
Do you have a preference for a default value? I know @gylaz supported the approach implemented in this PR because knowing the linter responsible for each file review had reporting value. For all file reviews created before this PR gets merged, we won't necessarily have that information. With reporting in mind should the default value be something like "unknown" or "default?"

teoljungberg replied:
For the default value, I don't have any preference. Thoughts @gylaz ?

gylaz replied:
Ideally, we would figure out what linter it should be. What if the migration set the linter name based on the file extension?

jbusser replied:
I could do that, but we would still need to decide on a rule for .js files, wouldn't we? If I'm given only a filename from a file review, I don't think it's possible to see what Hound configuration was in effect for a given repo during a given build, is it? I wouldn't necessarily know whether JShint, JSCS, or ESlint was active, right?

gylaz replied:
You are correct. I think at the time of the migration we could make a best guess -- say all the `*.js` are `jshint` since both `jscs` and `eslint` support were added recently and if we wrongly guess it's not a big deal. Alternatively, we can set `*.js` reviews to be `javascript` (and treat it as legacy before we made the switch to multi JS linters).

gylaz commented on app/models/file_review.rb:
```diff
@@ -2,6 +2,8 @@ class FileReview < ActiveRecord::Base
   belongs_to :build
   has_many :violations
 
+  validates_presence_of :linter_name
```
IMO, adding model-level validations is only useful when a user/client (UI or API) will consume them. If there is no such user, then DB constraints should be the gospel.

gylaz commented on app/models/linter/coffee_script.rb:
```diff
@@ -6,7 +6,9 @@ class CoffeeScript < Base
-      FileReview.create!(filename: commit_file.filename) do |file_review|
+      FileReview.create!(
+        filename: commit_file.filename,
+        linter_name: name) do |file_review|
```
`) do |file_review|` should be on its own line according to our guides.

jbusser commented:
Hi, guys --

I've fixed all the style errors; added the linter_name param to all Resque payloads; made a best effort at backfilling linter names; and I cleaned up that kind of gross implementation of `CompleteFileReview#file_review`.

Thank you for taking the time to review this.

jbusser commented:
Could I get a status update on this pull? I would like to get this committed since my company pays for Hound to lint some private repos. This change addresses a real problem we experience.

The CI error is the result of a failed security audit and does not appear to be related to this change. However, I'm happy to update the Administrate gem in a separate pull if you would like.

gylaz commented on db/migrate/20160314122828_add_linter_name_to_file_reviews.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  def backfill_linter_names
+    FileReview.reset_column_information
```
Why is this line needed?

gylaz commented on db/migrate/20160314122828_add_linter_name_to_file_reviews.rb:
```diff
@@ -0,0 +1,42 @@
+      Linter::Scss,
+      Linter::Swift,
+      Linter::Unsupported,
+    ]
```
When one of these linters is removed, this migration will fail for future contributors. Not sure the best way to handle this situation, maybe a one off task is a better fit for this?

teoljungberg replied:
Couldn't this use the list of linters already available in `Linter::Collection::LINTERS`?

jbusser replied:
@teoljungberg I certainly could. I took this collection from `Linter::Collection::LINTERS`. The only difference was removing the entry for `Linter::JSCS`. I chose to do that for data-accuracy. The regexp for matching the JSHint linter is similar to JSCS', but JSCS appears before JSHint in the array. That means every file with a `.js` suffix would be retroactively assigned a JSCS linter. Since jshint support predated JSCS support, I thought that would provide a more accurate history.

teoljungberg replied:
The issue that both @gylaz and I've pointed out is that this migration will have to be updated every time the list of linters changes, which is sub-optimal.

jbusser replied:
Yep. I get it. The migration depends too heavily on the state of the codebase. Related question: what is your opinion on referring AR models in migrations?

In the weeks since I pushed this code, I encountered a problem in my day job were we made a material change to a model that resulted in breaking earlier migrations. To correct this, we followed Justin Searls' advice outlined here: http://blog.testdouble.com/posts/2014-11-04-healthy-migration-habits.html

I still feel a migration is the only way out of this problem. I can't see how to ensure a not-null integrity constraint on file_reviews.linter_name with a combination of migration and one-off rake task.

So what if I:
* Defined a `MigrationFileReview` model here, breaking any dependency on `FileReivew`
* Defined a map of regexps and linter names, breaking the dependency on anything in `Linter::Collection`

gylaz replied:
> The only difference was removing the entry for Linter::JSCS.

Once #1119 lands, it won't be an issue anymore?

> what is your opinion on referring AR models in migrations?

We typically avoid using AR models in migrations, preferring SQL queries.
However, sometimes it will make sense to use an AR model, so what we woud do is:
```ruby
class AddLinterNameToFileReviews < ActiveRecord::Migration
  class FileReview < ActiveRecord::Base
  ...
end
```
Then `FileReview.find_each` will always work (well, until AR decides to rename/remove that method).

I think that in this case we can write a SQL query, with a `CASE` statement that does a regex match on the filename. It's more work, but it does not ties us to any additional dependencies.

jbusser replied:
:+1:

gylaz commented:
@jbusser The JSCS support is in beta, and now that it is merging with ESLint we are likely to drop support for it soon.

@teoljungberg @thorncp Since this situation won't arise if we drop support for JSCS, do we still think it's a good to do this? (analytics, future linters, etc.)

teoljungberg commented:
I can still see that there's value in keeping track which linter is responsible for which file review, eventhough `JSCS` is being moved out of the picture in the near future.

gylaz commented on db/migrate/20160314122828_add_linter_name_to_file_reviews.rb:
```diff
@@ -0,0 +1,42 @@
+      detect { |linter| linter.can_lint?(filename) }.
+      name.
+      demodulize.
+      underscore
```
Can we use the `Linter::Base#name` method here instead?

gylaz commented on spec/factories.rb:
```diff
@@ -15,6 +15,7 @@
     build
 
     filename "the_thing.rb"
+    linter_name "base"
```
Maybe this should be a real linter, like `ruby`? I don't think it can ever be `base`.

gylaz commented:
@jbusser A couple more things, then we can move forward with merging this. Thanks for pushing on this!

jbusser commented:
Thank you, @gylaz and @teoljungberg. I'll try to get those changes pushed up tonight.

jbusser commented:
I just wanted to let you guys know that this is ready for another review. The Ci failure is (still) the result of the out-of-date administrate gem and not necessarily related to this change. I'm still happy to fix that in separate pull if that would help.

teoljungberg commented:
@jbusser Try rebasing against master to remove the CI failure.

gylaz commented on db/migrate/20160430171015_backfill_linter_names_add_null_constraint.rb:
```diff
@@ -0,0 +1,35 @@
+        END;
+    SQL
+
+    ActiveRecord::Base.connection.execute(sql)
```
This could just be `execute(sql)`.

jbusser replied:
Would you prefer I change it? I used examples from `20150213132248_backfill_timestamps_add_null_constraints.rb` and `20141219224840_inactivate_repos_without_privacy_or_org_info.rb`. It's no issue for me either way.

gylaz commented on db/migrate/20160430171015_backfill_linter_names_add_null_constraint.rb:
```diff
@@ -0,0 +1,35 @@
+        WHEN filename ~* '\\.coffee(\\.js)?(\\.erb)?\\Z' THEN 'coffee_script'
+        WHEN filename ~* '\\.go\\Z' THEN 'go'
+        WHEN filename ~* '\\.haml\\Z' THEN 'haml'
+        WHEN filename ~* '(\\.es6|\\.es6\\.js)\\Z' THEN 'eslint'
```
Should thre be an `|` after `.es6`?

jbusser replied:
I believe we do want the `|` because we want to duplicate the regexp from `Linter::Eslint`:
```
    FILE_REGEXP = /.+(\.js|\.es6|\.es6\.js)\z/
```
I am deliberately skipping the `.js` pattern because I want that to fall through to match the jshint linter in the next line.

gylaz replied:
I see.

Should we move the `jshint` case above this? Right now, `jshint` will pickup `.es6.js` files (unless it's explicitly disabled). Or, are you making a best guess that if the file extension is `.es6.js` then it probably was also linted with ESLint explicitly?

jbusser replied:
Yes, the latter. It was an attempt at trying to make the historical record of linters and jobs associated with them line up with the order in which Hound supported them. Jshint support ESlint, so I wanted to give it the win, if you will, in the event of any ties.

thorncp commented:
This is great, thanks @jbusser! :+1: :dancers:

gylaz commented on db/migrate/20160430171015_backfill_linter_names_add_null_constraint.rb:
```diff
@@ -0,0 +1,35 @@
+        WHEN filename ~* '\\.py\\Z' THEN 'python'
+        WHEN filename ~* '\\.rb\\Z' THEN 'ruby'
+        WHEN filename ~* '\\.scss\\Z' THEN 'scss'
+        WHEN filename ~* '\\.swift\\Z' THEN 'swift'
```
I believe all of these need to be written with a single `\` as an escaping character.

jbusser replied:
I thought so, too, because of the heredoc, but it didn't work when I tested the migration. I needed the double escapes.

gylaz commented:
Thanks for your help and patience @jbusser!

jbusser commented:
This was my first significant open-source contribution. Thanks for making it a pleasant experience.

croaky commented:
Thank YOU, @jbusser!
